File: src/main/java/io/github/svegon/capi/event/render/ShouldDrawSideCallback.java
package io.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.block.BlockState;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.world.BlockView;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface ShouldDrawSideCallback {
    Event<ShouldDrawSideCallback> EVENT = EventFactory.createArrayBacked(ShouldDrawSideCallback.class, (state, world,
                pos, side, otherPos, callback) -> {}, listeners -> (state, world, pos, side, neighbor, callback) -> {
        for (ShouldDrawSideCallback listener : listeners) {
            listener.shouldDrawSide(state, world, pos, side, neighbor, callback);
        }
    });

    void shouldDrawSide(BlockState state, BlockView world, BlockPos pos, Direction side, BlockPos neighbor,
                        CallbackInfoReturnable<Boolean> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/ShouldDrawSideCallback.kt
package io.github.svegon.capi.event.render

import io.github.svegon.capi.event.render.ShouldDrawSideCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.block.BlockState
import net.minecraft.util.math.*
import net.minecraft.world.BlockView
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.function.Function
import kotlin.Array
import kotlin.Boolean
import kotlin.invoke

interface ShouldDrawSideCallback {
    fun shouldDrawSide(
        state: BlockState?, world: BlockView?, pos: BlockPos?, side: Direction?, neighbor: BlockPos?,
        callback: CallbackInfoReturnable<Boolean?>?
    )

    companion object {
        @JvmField
        val EVENT: Event<ShouldDrawSideCallback?> = EventFactory.createArrayBacked(
            ShouldDrawSideCallback::class.java,
            ShouldDrawSideCallback { state: BlockState?, world: BlockView?, pos: BlockPos?, side: Direction?, otherPos: BlockPos?, callback: CallbackInfoReturnable<Boolean?>? -> },
            Function { listeners: Array<ShouldDrawSideCallback?>? ->
                ShouldDrawSideCallback { state: BlockState?, world: BlockView?, pos: BlockPos?, side: Direction?, neighbor: BlockPos?, callback: CallbackInfoReturnable<Boolean?>? ->
                    for (listener in listeners) {
                        listener.shouldDrawSide(state, world, pos, side, neighbor, callback)
                    }
                }
            })
    }
}
